crucible-mir-comp/src/Mir/Compositional/State.hs

```diff
@@ -0,0 +1,80 @@
+{-# Language ConstraintKinds #-}
+{-# Language ImplicitParams #-}
+{-# Language TypeFamilies #-}
+module Mir.Compositional.State where
+
+import Control.Monad(foldM)
+import qualified Data.ByteString as BS
+import Data.IORef
+import Data.Set(Set)
+import qualified Data.Set as Set
+import Data.Text(Text)
+import qualified Data.Text as Text
+
+import qualified SAWCentral.Builtins as SAW
+import qualified SAWCore.SharedTerm as SAW
+import qualified CryptolSAWCore.CryptolEnv as SAW
+import qualified CryptolSAWCore.Prelude as SAW
+import qualified SAWCoreWhat4.What4 as SAW
+import qualified SAWCoreWhat4.ReturnTrip as SAW
+import qualified What4.Expr as W4
+
+
+data MirState sym = MirState {
+  mirSharedContext   :: SAW.SharedContext,
+  -- ^ Shared context for building Cryptol terms
+
+  mirCryEnv           :: IORef SAW.CryptolEnv,
+  -- ^ Inforamtion about what terms are loaded
+
+  mirKeepUninterp     :: IORef (Set Text),
+  -- ^ Set of names we'd like to keep uninterpreted;
+  -- we use `_cryEnv` to compute what they refer to.
+
+  mirUninterpFunCache :: IORef (SAW.SymFnCache sym),
+  -- ^ A cache used by translation to What4, to keep track of
+  -- which uninterpred functions we've already made.
+
+  mirSAWCoreState :: SAW.SAWCoreState (SymScope sym)
+}
+
+type UsesMirState sym = (?mirState :: MirState sym)
+
+type family SymScope sym
+type instance SymScope (W4.ExprBuilder n st fs) = n
```

Comment:
If I understand correctly, all of the code in `crux-mir-comp` assumes a `sym ~ ExprBuilder t st fs` constraint anyway, so I wonder if it would be more direct to just parameterized `MirState` by the `t` type parameter and monomorphisze `sym` to `Sym`.